### PR TITLE
feat(HF): simulating repayment-borrow actions and exposing the healthFactor

### PR DESCRIPTION
**Pull request by tomrpl**
Here is the first iteration of the virtual repayment/borrow.

With Merlin, we discussed abut the fact that having assets only in input is enough in the current requirement from the integrator that requested it.

Would you mind spend some time thinking about potential hedge case that would break the computation? 

Tried to cover some in our test.

To get them pass, I had to bound the borrowedAmount with the MAX_AMOUNT being MAX_TEST_AMOUNT/2 instead of letting to MAX_TEST_AMOUNT.

Do you have a workaround in mind @Rubilmax ?

Thank you guys!

### CODE REVIEW

File: src/morpho-blue/MorphoBlueSnippets.sol [hunk 1]
```diff
@@ -146,6 +146,54 @@ contract MorphoBlueSnippets {
...
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        // Prevent overflow
+        uint256 newBorrowed = borrowed + borrowAmount;
+        require(newBorrowed >= borrowed, "Borrow amount overflow");
+
+        uint256 maxBorrow = collateral.mulDivDown(collateralPrice, ORACLE_PRICE_SCALE).wMulDown(marketParams.lltv);
+
+        return newBorrowed == 0 ? type(uint256).max : maxBorrow.wDivDown(newBorrowed);
+    }
```
Comment (MerlinEgalite) [inline src/morpho-blue/MorphoBlueSnippets.sol:196]: I would extract the logic into another contract tbh as we don't care about the rest of the logic

File: src/morpho-blue/MorphoBlueSnippets.sol [hunk 2]
```diff
@@ -146,6 +146,54 @@ contract MorphoBlueSnippets {
...
+        Id id,
+        address user,
+        uint256 repaymentAmount
+    ) public view returns (uint256 healthFactor) {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        // Ensure repaymentAmount doesn't exceed the borrowed amount
+        uint256 newBorrowed = borrowed > repaymentAmount ? borrowed - repaymentAmount : 0;
```
Comment (MerlinEgalite) [inline src/morpho-blue/MorphoBlueSnippets.sol:166]: I would revert in this case instead (safer)

File: src/morpho-blue/MorphoBlueSnippets.sol [hunk 3]
```diff
@@ -146,6 +146,54 @@ contract MorphoBlueSnippets {
...
+        address user,
+        uint256 borrowAmount
+    ) public view returns (uint256 healthFactor) {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        // Prevent overflow
+        uint256 newBorrowed = borrowed + borrowAmount;
+        require(newBorrowed >= borrowed, "Borrow amount overflow");
```
Comment (MerlinEgalite) [inline src/morpho-blue/MorphoBlueSnippets.sol:191]: There's no overflow possible as we're using solidity 0.8.x

If `borrowed + borrowAmount` is too large it will revert

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    IMorpho public immutable morpho;
+
+    /* CONSTRUCTOR */
+
+    /// @notice Constructs the contract.
+    /// @param morphoAddress The address of the Morpho Blue contract.
+    constructor(address morphoAddress) {
+        require(morphoAddress != address(0), "Invalid Morpho address");
+        morpho = IMorpho(morphoAddress);
+    }
```
[the remark this answers is not in the document]
Comment (tomrpl) [reply src/morpho-blue/VirtualHealthFactor.sol:33]: I think that is not necessary. 
I'll let that open for Romain to see it, no strong opinion on my side though

[the remark this answers is not in the document]
Comment (Rubilmax) [reply src/morpho-blue/VirtualHealthFactor.sol:33]: Not either, snippets are here to should how to interact with it imo not how to store information

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        healthFactor = maxBorrow.wDivDown(borrowed);
+    }
+
+    /// @notice Calculates the health factor of a user after a virtual repayment.
+    /// @param marketParams The parameters of the market.
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @param repaymentAmount The amount of assets to be virtually repaid.
+    /// @return healthFactor The calculated health factor after the virtual repayment.
+    function userHealthFactorAfterVirtualRepayment(
```
Comment (Rubilmax) [inline src/morpho-blue/VirtualHealthFactor.sol:63]:
```suggestion
    function userHypotheticalHealthFactor(
```

just a suggestion, current naming is expressive neough

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        MarketParams memory marketParams,
+        Id id,
+        address user,
+        uint256 repaymentAmount
```
Comment (Rubilmax) [inline src/morpho-blue/VirtualHealthFactor.sol:67]:
```suggestion
        uint256 repaidAssets
```

to keep naming consistent with Morpho Blue

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    /// @notice Calculates the health factor of a user in a specific market.
+    /// @param marketParams The parameters of the market.
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @return healthFactor The calculated health factor.
+    function userHealthFactor(MarketParams memory marketParams, Id id, address user)
+        public
+        view
+        returns (uint256 healthFactor)
```
Comment (Rubilmax) [inline src/morpho-blue/VirtualHealthFactor.sol:45]:
```suggestion
        returns (uint256)
```

Return statements are clearer but idrc

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @param repaymentAmount The amount of assets to be virtually repaid.
+    /// @return healthFactor The calculated health factor after the virtual repayment.
+    function userHealthFactorAfterVirtualRepayment(
+        MarketParams memory marketParams,
+        Id id,
+        address user,
+        uint256 repaymentAmount
+    ) public view returns (uint256 healthFactor) {
```
Comment (Rubilmax) [inline src/morpho-blue/VirtualHealthFactor.sol:68]:
```suggestion
    ) public view returns (uint256) {
```

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        // Revert if repaymentAmount exceeds the borrowed amount
+        require(repaymentAmount <= borrowed, "Repayment amount exceeds borrowed amount");
```
Comment (Rubilmax) [inline src/morpho-blue/VirtualHealthFactor.sol:74]: We could simply let the snippet revert with underflow just like in Morpho Blue (idrc)

File: src/morpho-blue/VirtualHealthFactor.sol [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        MarketParams memory marketParams,
+        Id id,
+        address user,
+        uint256 borrowAmount
+    ) public view returns (uint256 healthFactor) {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        uint256 newBorrowed = borrowed + borrowAmount;
```
Comment (Rubilmax) [inline src/morpho-blue/VirtualHealthFactor.sol:98]: Note here you can overflow for a similar reason than during repay. Yet we don't revert with a dedicated message in this case (inconsistency). I would prefer not reverting during repay

File: src/morpho-blue/VirtualHealthFactor.sol (hunk 5 above)
Comment (Rubilmax) [reply src/morpho-blue/VirtualHealthFactor.sol:63]:
```suggestion
    function userHealthFactorAfterVirtualRepay(
```

This is more consistent with the borrow naming

**Review by Rubilmax — APPROVED** [review]
Looks good

Take a look at: https://morpholabs.slack.com/archives/C02P3QE2Q8G/p1725447118759289

You have no new line at EOF in tests (hence the 🚫 symbol in github reviews)
